ClearlyEnergy/BEAM-OPT
# enhancement: Decrease Execution time of Optimization

#### What's this PR do?

There were multiple redundant and unused computations in `_optimize`. Most of these computations have been identified and reduced or eliminated.

The optimizer also used pandas Series objects for doing some computations. These computations were converted to use the underlying numpy arrays which significantly reduces computation time for vector operations.

Before these changes, [this analysis on DC Staging](https://dc-staging.beam-portal.org/app/#/analyses/194/runs/640) took nearly 8 minutes. Before changes were made, I benchmarked this analysis on my system and benchmarked 4 minutes. After the changes made here, the entire computation takes approximately 1 minute to complete on my system. The result of the computation appears identical (optimizer currently has no programmatic tests).

#### Manual Testing

Do the following before applying the changes made by this PR to get a benchmark for execution time.
1. Copy the [Emission Scenario](https://dc-staging.beam-portal.org/app/#/accounts/359/emission_scenarios/9) used for the previously mentioned analysis.
2.  Upload the following Measures/Scenarios to a test building [BEAMopt_bldg3.xlsx](https://github.com/ClearlyEnergy/BEAM-OPT/files/14089101/BEAMopt_bldg3.xlsx)
3. Copy the analysis input (budget per year, selected emission scenario) used in the analysis (linked above).
4. Run the analysis and note the time to complete.
5. Apply the changes from this PR and run the same analysis again. Note difference in time. Ensure that test output is same (same Scenarios selected, reduced emissions, etc)

#### What are the relevant tickets?

https://clearlyenergy-inc.monday.com/boards/5290716771/pulses/5839414191
